tree-diff

```diff
@@ -0,0 +1,150 @@
+        props["mode"] = [stat1.st_mode, stat2.st_mode]
+        return False
+    if stat1.st_uid != stat2.st_uid:
+        props = differences.setdefault(path, {})
+        props["uid"] = [stat1.st_uid, stat2.st_uid]
+    if stat1.st_gid != stat2.st_gid:
+        props = differences.setdefault(path, {})
+        props["gid"] = [stat1.st_gid, stat2.st_gid]
+    if stat1.st_size != stat2.st_size:
+        props = differences.setdefault(path, {})
+        props["size"] = [stat1.st_size, stat2.st_size]
+    return True
+
+
+def selinux_diff(path1, path2, path, differences):
+    try:
+        label1 = os.getxattr(path1, b"security.selinux", follow_symlinks=False).decode()
+        label2 = os.getxattr(path2, b"security.selinux", follow_symlinks=False).decode()
+    except OSError:
+        return True
+    if label1 != label2:
+        props = differences.setdefault(path, {})
+        props["selinux"] = [os.fsdecode(label1).encode(), os.fsdecode(label2).encode()]
+        return False
+    return True
+
+
+def content_diff(name, dir_fd1, dir_fd2, path, differences):
+    try:
+        fd1 = os.open(name, flags=os.O_RDONLY, dir_fd=dir_fd1)
+    except OSError:
+        return
+    try:
+        fd2 = os.open(name, flags=os.O_RDONLY, dir_fd=dir_fd2)
+    except OSError:
+        os.close(fd1)
+        return
+    try:
+        for (byte_block1, byte_block2) in zip(iter(lambda f=fd1: os.read(f, 4096), b""),
+                                              iter(lambda f=fd2: os.read(f, 4096), b"")):
+            if byte_block1 != byte_block2:
+                props = differences.setdefault(path, {})
+                props["content"] = "different"
+                break
+    finally:
+        os.close(fd1)
+        os.close(fd2)
+
+
+def symlink_diff(name, dir_fd1, dir_fd2, path, differences):
+    try:
+        target1 = os.readlink(name, dir_fd=dir_fd1)
+        target2 = os.readlink(name, dir_fd=dir_fd2)
+    except OSError:
+        return
+    if target1 != target2:
+        props = differences.setdefault(path, {})
+        props["symlink"] = [os.fsdecode(target1).encode(), os.fsdecode(target2).encode()]
+
+
+def diff_aux(dir_fd1, dir_fd2, path, report):
+    entries1 = set()
+    with os.scandir(dir_fd1) as it:
+        for dirent in it:
+            try:
+                stat2 = os.stat(dirent.name, dir_fd=dir_fd2, follow_symlinks=False)
+            except FileNotFoundError:
+                report["deleted_files"] += [os.path.join(path, dirent.name)]
+                continue
+            entries1.add(dirent.name)
+            stat1 = dirent.stat(follow_symlinks=False)
+            selinux_diff(os.path.join(f"/proc/self/fd/{dir_fd1}", dirent.name),
+                         os.path.join(f"/proc/self/fd/{dir_fd2}", dirent.name),
+                         os.path.join(path, dirent.name),
+                         report["differences"])
+            if not stat_diff(stat1, stat2, path, report["differences"]):
```

Comment:
I guess this is just overlook: replace `path` with `os.path.join(path, dirent.name)` to list entries instead of their directories

Reply:
Good catch! Thanks!